sycl/doc/extensions/proposed/sycl_ext_oneapi_kernel_properties_macro.asciidoc

```diff
@@ -0,0 +1,329 @@
+  __kernel_properties(properties{streaming_interface_accept_downstream_stall,
+                                  pipelined<1>})
+  void operator()() const {
+    ... // kernel body
+  }
+};
+```
+
+== Alternatives considered
+
+_This is a non normative section and is only present here to aid discussion.
+This will be removed before the extension is published_
+
+=== Inheriting from a base functor with properties applied
+
+This would look something like:
+```c++
+using namespace sycl::ext::intel::experimental;
+using namespace sycl::ext::oneapi::experimental;
+
+// *****************************************************************************
+// All of this will be in the headers
+using streaming_interface_t = streaming_interface_key::value_t<
+    streaming_interface_options_enum::accept_downstream_stall>;
+
+template <int pipeline_directive_and_initiation_interval = -1>
+using pipelined_t =
+    pipelined_key::value_t<pipeline_directive_and_initiation_interval>;
+
+template <typename... Props> struct KernelPropertiesBase {
+  using ptype = ext::oneapi::experimental::detail::properties_t<Props...>;
+  auto get(properties_tag) { return ptype{}; }
+};
+// **************************** End of header **********************************
+
+// **** user code **********
+struct MyKernel : KernelPropertiesBase<decltype(properties{
+  streaming_interface_accept_downstream_stall,
+  pipelined})> {
+
+public:
+  annotated_arg<int *, ...> arg_a;
+  int n;
+  void operator()() const { ... }
+};
+```
+
+Issues with this approach:
+
+- Aggregate initialization requires that an additional empty
+`std::initializer_list` be passed:
+
+```c++
+auto* array_a = sycl::malloc_host(...);
+int N = ...;
+q.single_task(MyKernel{{}, array_A, N}).wait();
+```
+
+- An additional kernel argument is generated which is unexpected. This will
+require the compiler to chop it off and that in turn would require additional
+changes to other dependent parts of the compiler like the runtime.
+
+Due to these issues this option was discarded.
+
+=== Defining properties into a type in the functor
+
+The following pseudocode shows this approach:
+
+```c++
+struct MyKernel {
+  // set the properties
+  using kernel_props_type = decltype(properties{
+      streaming_interface_accept_downstream_stall,
+    pipelined})
+  
+  ... // kernel arguments
+  void operator()() const { ... }
+};
+```
+
+The problem with this approach is that it will only work with compile time
+constant properties whereas the `get` method was designed to work with runtime
+properties as well. Hence, if we use this approach then we have 2 APIs for
+setting properties that are not equivalent.
+
+This approach potentially requires non trivil amount of updates to the
+implementation to keep the 2 APIs working.
+
+Due to these issues this option was discarded.
+
+== Issues
+
+1. Name of the macro.
+2. Style of the macro.
+3. Do we recommend using `macro();` syntax or `macro() void operator()() {...}`
+syntax.
```

Comment:
I don't understand this issue #3.  What is the choice to decide?

Reply:
Hey Greg, I updated the issue to explain what I meant. The context is that we are trying to figure out BKM for very novice users for the FPGA compiler product. Perhaps we do not need to discuss that here a we can provide our own BKMs in FPGA docs however I added it here in case it makes sense for general oneapi users too.